x/subscription/keeper/cu_tracker.go

```diff
@@ -5,34 +5,188 @@ import (
+	key := types.CuTrackerKey(sub, info.provider, info.chainID)
+	if shouldRemove {
+		return k.cuTrackerFS.DelEntry(ctx, key, uint64(ctx.BlockHeight()))
+	}
+
+	var trackedCu types.TrackedCu
+	found := k.cuTrackerFS.FindEntry(ctx, key, info.block, &trackedCu)
+	if !found {
+		return utils.LavaFormatError("cannot reset CU tracker", fmt.Errorf("trackedCU entry is not found"),
+			utils.Attribute{Key: "sub", Value: sub},
+			utils.Attribute{Key: "provider", Value: info.provider},
+			utils.Attribute{Key: "chainID", Value: info.chainID},
+			utils.Attribute{Key: "block", Value: strconv.FormatUint(info.block, 10)},
+		)
+	}
+
+	trackedCu.Cu = 0
+	k.cuTrackerFS.ModifyEntry(ctx, key, info.block, &trackedCu)
+	return nil
+}
+
+type trackedCuInfo struct {
+	provider  string
+	chainID   string
+	trackedCu uint64
+	block     uint64
+}
+
+func (k Keeper) getSubTrackedCuInfo(ctx sdk.Context, sub string) (trackedCuList []trackedCuInfo, totalCuUsedBySub uint64) {
+	keys := k.GetAllSubTrackedCuIndices(ctx, sub)
+
+	for _, key := range keys {
+		_, provider, chainID := types.DecodeCuTrackerKey(key)
+		cu, entryBlock, found, _ := k.GetTrackedCu(ctx, sub, provider, chainID)
+		if !found {
+			utils.LavaFormatWarning("cannot remove cu tracker", legacyerrors.ErrKeyNotFound,
+				utils.Attribute{Key: "sub", Value: sub},
+				utils.Attribute{Key: "provider", Value: provider},
+				utils.Attribute{Key: "block", Value: strconv.FormatUint(entryBlock, 10)},
+			)
+		}
+		trackedCuList = append(trackedCuList, trackedCuInfo{
+			provider:  provider,
+			trackedCu: cu,
+			chainID:   chainID,
+			block:     entryBlock,
+		})
+		totalCuUsedBySub += cu
+	}
+
+	return trackedCuList, totalCuUsedBySub
+}
+
+// remove only before the sub is deleted
+func (k Keeper) RewardAndResetCuTracker(ctx sdk.Context, cuTrackerTimerKeyBytes []byte) {
+	sub, shouldRemove := types.DecodeCuTrackerTimerKey(string(cuTrackerTimerKeyBytes))
+	trackedCuList, totalCuUsedBySub := k.getSubTrackedCuInfo(ctx, sub)
+
+	for _, trackedCuInfo := range trackedCuList {
+		trackedCu := trackedCuInfo.trackedCu
+		provider := trackedCuInfo.provider
+		chainID := trackedCuInfo.chainID
+		block := trackedCuInfo.block
+
+		plan, err := k.GetPlanFromSubscription(ctx, sub, block)
+		if err != nil {
+			utils.LavaFormatError("cannot find subscription's plan", types.ErrCuTrackerPayoutFailed,
+				utils.Attribute{Key: "sub_consumer", Value: sub},
+			)
+			return
+		}
+
+		err = k.resetCuTracker(ctx, sub, trackedCuInfo, shouldRemove)
+		if err != nil {
+			utils.LavaFormatError("removing/reseting tracked CU entry failed", err,
+				utils.Attribute{Key: "provider", Value: provider},
+				utils.Attribute{Key: "tracked_cu", Value: trackedCu},
+				utils.Attribute{Key: "chain_id", Value: chainID},
+				utils.Attribute{Key: "sub", Value: sub},
+				utils.Attribute{Key: "block", Value: ctx.BlockHeight()},
+			)
+			return
+		}
+
+		// TODO: deal with the reward's remainder (uint division...)
+		if trackedCu > totalCuUsedBySub {
```

Comment:
this will never happen since totalCuUsedBySub is calculated from sums of trackedCu

Reply:
done. see https://github.com/lavanet/lava/commit/36c4bd192408759547b2588605992db3d3b23f21